### PR TITLE
Scaling Example NF Updates

### PR DESCRIPTION
**Pull request by catherinemeadows**
Modified the Scaling Example NF to assign spawned children with the same service ID as the parent and updated the README with more detail about callback and advanced rings modes. 

- addressed issue [250](https://github.com/sdnfv/openNetVM/issues/250)

## Summary:

**Usage:**

| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          | [250](https://github.com/sdnfv/openNetVM/issues/250)
| Breaking API changes     |
| Internal API changes     |
| Usability improvements   |  
| Bug fixes                |
| New functionality        |
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 

## Merging notes:
 - Dependencies: None  

**TODO before merging :**
 - [X] PR is ready for review

## Test Plan:
- tested NF in a chain (speed tester -> scaling -> basic monitor) to verify correct packet flow

### CODE REVIEW

**Comment by twood02** [comment]
@catherinemeadows 
Can you remind me -- Is the idea of this NF that it spawns N child NFs all with the same service ID as the "main" NF that started initially. Packets are then distributed across the `N+1` NFs.  All of these NFs then forward packets to whatever destination ID was set on the command line.

Is this right?  Or does the main NF not actually have a run loop so there are `N` nfs total receiving packets (only the children)?

**Comment by catherinemeadows** [comment]
@twood02 Yes, the parent NF spawns N children NFs with the same service ID and the packets are distributed across the N+1 NFs and forwarded to the destination ID. Should I add this to the documentation?

File: examples/scaling_example/README.md [hunk 1]
```diff
@@ -3,18 +3,18 @@ Scaling Example NF
 This program showcases how to use NF scaling. It shows how to use the ONVM scaling API or use the advanced rings mode and implement your own scaling logic.
 The NF will spawn X children (provided by `-n`, defaulting to 1 if arg isn't provided) and send packets to destination NF.
```
Comment (twood02) [inline examples/scaling_example/README.md]: Revise this line with longer description of N+1 nfs

File: examples/scaling_example/README.md [hunk 2]
```diff
@@ -3,18 +3,18 @@ Scaling Example NF
...
+ - **Advanced Rings Mode**: This mode will be used when the advanced rings flag, `-a`, is set. The NF will do its own pthread creation and manage the NF rx/tx ring on its own. In advanced rings mode, the default action for processing packets in `thread_main_loop` is to forward packets to the destination NF provided by `-d`. 
 
 Compilation and Execution
 --
 ```
 cd examples
 make
 cd scaling_example
-./go.sh SERVICE_ID -d DST_ID [ADVANCED_RINGS]
+./go.sh SERVICE_ID -d DST_ID [NUM_CHILDREN] [ADVANCED_RINGS]
```
Comment (twood02) [inline examples/scaling_example/README.md]: does the `go` script hide the -n and -a arguments or is that needed here? Maybe this should be:

```
 ./go.sh SERVICE_ID -d DST_ID [-n NUM_CHILDREN] [-a]
```

**Review by twood02 — CHANGES_REQUESTED** [review]

**Review by twood02 — APPROVED** [review]
looks good to me

**Comment by twood02** [comment]
@NoahChinitzGWU can you build and run this code on one of your systems to verify it is working as expected?  Post a screenshot/paste the output here

**Comment by NoahChinitz** [comment]
Manager:
Screenshot from 2022-03-21 17-32-06

Scaling NF Left Terminal, Monitor NF Right Terminal:
Screenshot from 2022-03-21 17-32-30

Does this seem correct, or should I be seeing 3 additional scaling NFs in the manager?

**Comment by catherinemeadows** [comment]
They should scale once you send packets to it (you can use speed tester)

**Comment by NoahChinitz** [comment]
Manager:
Screenshot from 2022-03-21 17-44-17

Works!

**Comment by NoahChinitz** [comment]
Screenshot from 2022-03-21 17-45-01

NFs used (scaling left, basic monitor middle, speed tester right)

**Review by NoahChinitz — APPROVED** [review]
See my previous comments for screenshots of working output

**Comment by twood02** [comment]
@dennisafa - can you do a very quick review and approve this if nothing jumps out at you?

**Review by dennisafa — APPROVED** [review]
Changes look good to me. Unrelated to the bug itself, I think it would make sense to do the scaling differently. Right now, we spawn the child NF's as soon as we receive a packet in the packet_handler_with_scaling() function. The new NF's packet handler is then set to packet_handler_fwd(). Why not spawn all of the child NF's upon startup, rather than wait for the first packet? The current implementation would make more sense if we were spawning the child NF's based on say, the packet count/rate.

**Pull request merged**